### PR TITLE
Bluetooth: Controller: Fix to stop Extended Auxiliary Scan context

### PR DESCRIPTION
**Pull request by cvinayak**
Update ull_disable implementation to return -EALREADY if
LLL event is already disabled.

Fix any stray Extended Auxiliary PDU from being scanned
when disabling Extended Scanning.

Updated Extended Scan disable implementation to find any
active auxiliary scan context and stop them.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

### CODE REVIEW

**Review by kruithofa — APPROVED** [review]

File: subsys/bluetooth/controller/ll_sw/ull.c [hunk 1]
```diff
@@ -1722,16 +1723,17 @@ int ull_ticker_stop_with_mark(uint8_t ticker_handle, void *param,
...
 
 	mark = ull_disable_unmark(param);
 	if (mark != param) {
 		return -ENOLCK;
 	}
 
+	if (err && (err != -EALREADY)) {
+		return err;
+	}
+
```
Comment (Thalley) [inline subsys/bluetooth/controller/ll_sw/ull.c:1736]: In this change you've change the call to `ull_disable_unmark` to be done onconditionally as the only place in the code base. Is that intentional? 

Later I see that you do a similar change where it is still called if there's an error, but only if the error is `EALREADY`. Should that have been done here as well?

**Review by Thalley — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/controller/ll_sw/ull_scan_aux.c [hunk 2]
```diff
@@ -926,15 +957,37 @@ int ull_scan_aux_stop(struct ll_scan_aux_set *aux)
...
-			return -EBUSY;
+		err = ull_disable(&aux->lll);
+		if (err) {
+			return err;
 		}
 
 		mfy.fp = flush;
+
+	} else if (!IS_ENABLED(CONFIG_BT_CTLR_SYNC_PERIODIC)) {
+		/* ULL scan auxiliary PDU reception scheduling stopped
```
Comment (ppryga) [inline subsys/bluetooth/controller/ll_sw/ull_scan_aux.c:968]: Just comment.

In my humble opinion, there should be a documentation that puts this kind of functions in all possible flows (e.g.  some diagrams) to make it easier to understand. This kind of functions depend on some global state and it is hard to put it into right context to get thorough understanding. Comments doesn't solve the problem. Comments are helpful for a person that is deep into the context.

File: subsys/bluetooth/controller/ll_sw/ull.c (hunk 1 above)
Comment (Thalley) [reply subsys/bluetooth/controller/ll_sw/ull.c:1736]: Resolved outside of Github :)

**Review by Thalley — APPROVED** [review]

File: subsys/bluetooth/controller/ll_sw/ull.c (hunk 1 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull.c:1736]: `ull_disable_mark` and `ull_disable_unmark` are used to save the current state/role context pointer when ticker instance is being stopped so that the ticker operation callback can detect a race when both thread and ULL execution context try to call `ticker_stop()`.

Here at line 1733, if `ull_disable()` returns `-EALREADY` then the error is returned back to the caller of `ull_ticker_stop_with_mark()`.

Whereas in line 2590, if `ull_disable()` returns `-EALREADY` then we ignore it as `disable()` which is the caller is directly called by HCI interface to disable the advertiser.

In other words,

In the first instance, `ull_ticker_stop_with_mark()` function is reused across the Controller and the error code is returned back to caller.

In the second instance, `disable()` function is direct call from HCI interface to disable advertising and error (except `-EALREADY` ) return from ull_disable() is design fault, hence use of LL_ASSERT

File: subsys/bluetooth/controller/ll_sw/ull_scan_aux.c [hunk 3]
```diff
@@ -747,13 +755,36 @@ void ull_scan_aux_done(struct node_rx_event_done *done)
...
+		}
+
 		/* Setup the disabled callback to flush the auxiliary PDUs */
 		hdr = &aux->ull;
 	}
 
-	LL_ASSERT(!hdr->disabled_cb);
-	hdr->disabled_param = aux;
-	hdr->disabled_cb = done_disabled_cb;
+	if (ull_ref_get(hdr) == 0U) {
```
[the remark this answers is not in the document]
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_scan_aux.c:772]: Scanning with interval and window will not have overlapping prepare, hence when the radio event is done (handle in separate `ull_done`mayfly) and `ull_scan_aux_done` (handle in `rx_demux` mayfly) they can be in any order of execution depending which runs first the reference count could be `0`. I.e., if `ull_done` runs before `ull_scan_aux_done`.

When under continuous scan, then usually at the end of the window the reference count would be > `1U`.

File: subsys/bluetooth/controller/ll_sw/ull_scan_aux.c [hunk 4]
```diff
@@ -747,13 +755,36 @@ void ull_scan_aux_done(struct node_rx_event_done *done)
...
+	}
+}
+
+struct ll_scan_aux_set *ull_scan_aux_set_get(uint8_t handle)
+{
+	if (handle >= CONFIG_BT_CTLR_SCAN_AUX_SET) {
+		return NULL;
+	}
+
+	return &ll_scan_aux_pool[handle];
```
[the remark this answers is not in the document]
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_scan_aux.c:787]: Caller `ull_scan_disable` checks if the context is associated with the scan context that is being disable, and that is a sufficient check for now.

File: subsys/bluetooth/controller/ll_sw/ull_scan.c [hunk 5]
```diff
@@ -462,13 +469,43 @@ uint8_t ull_scan_disable(uint8_t handle, struct ll_scan_set *scan)
...
+		aux_scan_lll = aux->parent;
+		if (!aux_scan_lll) {
+			continue;
+		}
+
+		aux_scan = HDR_LLL2ULL(aux_scan_lll);
+		if (aux_scan == scan) {
+			err = ull_scan_aux_stop(aux);
+			if (err && (err != -EALREADY)) {
+				return BT_HCI_ERR_CMD_DISALLOWED;
```
[the remark this answers is not in the document]
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_scan.c:503]: This conditional should not occur, should I replace this with a LL_ASSERT?

File: subsys/bluetooth/controller/ll_sw/ull_scan_aux.c (hunk 2 above)
Comment (carlescufi) [reply subsys/bluetooth/controller/ll_sw/ull_scan_aux.c:968]: I agree with this, but this is outside the scope of this PR. @cvinayak I believe already has plans to document the LL once the crunch with Audio is complete.

File: subsys/bluetooth/controller/ll_sw/ull_scan.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (ppryga) [reply subsys/bluetooth/controller/ll_sw/ull_scan.c:503]: If that is something that may not happen and it will get to some unexpected state then in my humble opinion we should assert. I would not create an assert if there is a way to recover from the situation. The change can be added in follow up PR.
